### PR TITLE
BUG: Use C call to sysctlbyname for AVX detection on MacOS

### LINKED ISSUE #7801: AVX detection fails on MacOS if $PATH doesn't contain /usr/sbin
Only on MacOS.  Installed 1.11.1 via pip on Python 3.5.2.

```
~ $ PATH=/usr/local/bin:/usr/bin:/bin python3 -c 'import numpy'
sh: sysctl: command not found
~ $ PATH=/usr/sbin:/usr/local/bin:/usr/bin:/bin python3 -c 'import numpy'
```

See: https://github.com/numpy/numpy/blob/1429c606643d1ad305e710c4a31cb6f398d04c53/numpy/_build_utils/src/apple_sgemv_fix.c#L69

Note that this uses `system()`, which is just plain silly, and explains why it fails if the environment variables are not what you expect.  Here is an alternative, self-contained program that works more reliably:

```
#include <stdio.h>
#include <sys/types.h>
#include <sys/sysctl.h>
int cpu_supports_avx(void) {
    int enabled, r;
    size_t length = sizeof(enabled);
    r = sysctlbyname("hw.optional.avx1_0", &enabled, &length, NULL, 0);
    return r == 0 && enabled != 0;
}
int main(int argc, char **argv) {
    printf("AVX supported: %d\n", cpu_supports_avx());
    return 0;
}
```

### PR DESCRIPTION
**Pull request by bran206**
Fixes #7801.

### CODE REVIEW

File: numpy/_build_utils/src/apple_sgemv_fix.c [hunk 1]
```diff
@@ -66,12 +69,21 @@ static int AVX_and_10_9 = 0;
 /* Dynamic check for AVX support
  * __builtin_cpu_supports("avx") is available in gcc 4.8,
  * but clang and icc do not currently support it. */
-#define cpu_supports_avx()\
-(system("sysctl -n machdep.cpu.features | grep -q AVX") == 0)
-
+#define cpu_supports_avx() ({\
```
Comment (eric-wieser) [inline numpy/_build_utils/src/apple_sgemv_fix.c]: Any reason not just to use a `static inline cpu_supports_avx(void)` function here? This syntax is not standard.

Comment (bran206) [reply numpy/_build_utils/src/apple_sgemv_fix.c]: Hey Eric,
Yeah I made a mistake with that, I was a bit too focused on making it a macro haha. Let me make some changes and merge it again.
Thanks for the input,
Bran

P.S please reject it.

**Comment by charris** [comment]
Just keep pushing changes to the same PR. When everything is fixed up, you can squash the commits.

File: numpy/_build_utils/src/apple_sgemv_fix.c [hunk 2]
```diff
@@ -66,12 +69,27 @@ static int AVX_and_10_9 = 0;
...
+	
 /* Check if we are using MacOS X version 10.9 */
-#define using_mavericks()\
-(system("sw_vers -productVersion | grep -q 10\\.9\\.") == 0)
+static inline int using_mavericks() {
+	int r;
+	char str[32] = {0};
+	size_t size = sizeof(str);
+	r = sysctlbyname("kern.osproductversion", str, &size, NULL, 0);
+	if ( r == 0 && strncmp(str, "10.9", strlen("10.9")) == 0)
```
Comment (eric-wieser) [inline numpy/_build_utils/src/apple_sgemv_fix.c]: What purpose does `strncmp` serve here vs just using `strcmp`? In general, `strncmp(a, b, strlen(b))` will never be different to `strcmp(a, b)`

**Comment by bran206** [comment]
Hi Eric,
It’s mostly to mitigate issues with subversions of 10.9 which is what the regular expression that was implemented before did.

File: numpy/_build_utils/src/apple_sgemv_fix.c (hunk 2 above)
Comment (eric-wieser) [reply numpy/_build_utils/src/apple_sgemv_fix.c]: My mistake this is correct

**Comment by eric-wieser** [comment]
@bran206: Gotcha, I get it now. Thanks for teaching me a new C idiom to look out for.

**Comment by bran206** [comment]
No worries mate.

**Comment by charris** [comment]
C style needs fixing, four spaces for indent, no tabs, etc. See `doc/C_STYLE_GUIDE.rst.txt`.

**Comment by bran206** [comment]
Done, i set the spaces, i don't see anything else i did wrong

File: numpy/_build_utils/src/apple_sgemv_fix.c [hunk 3]
```diff
@@ -66,12 +69,27 @@ static int AVX_and_10_9 = 0;
...
+        return 0;
+}
+	
 /* Check if we are using MacOS X version 10.9 */
-#define using_mavericks()\
-(system("sw_vers -productVersion | grep -q 10\\.9\\.") == 0)
+static inline int using_mavericks() {
+    int r;
+    char str[32] = {0};
+    size_t size = sizeof(str);
```
Comment (charris) [inline numpy/_build_utils/src/apple_sgemv_fix.c]: Blank line after.

File: numpy/_build_utils/src/apple_sgemv_fix.c [hunk 4]
```diff
@@ -66,12 +69,27 @@ static int AVX_and_10_9 = 0;
...
+    r = sysctlbyname("kern.osproductversion", str, &size, NULL, 0);
+    if ( r == 0 && strncmp(str, "10.9", strlen("10.9")) == 0)
+        return 1;
```
Comment (charris) [inline numpy/_build_utils/src/apple_sgemv_fix.c]: Use brackets even for single statement if blocks.

File: numpy/_build_utils/src/apple_sgemv_fix.c [hunk 5]
```diff
@@ -66,12 +69,27 @@ static int AVX_and_10_9 = 0;
...
+    if ( r == 0 && enabled != 0) 
+        return 1;
+    else
+        return 0;
+}
+	
 /* Check if we are using MacOS X version 10.9 */
-#define using_mavericks()\
-(system("sw_vers -productVersion | grep -q 10\\.9\\.") == 0)
+static inline int using_mavericks() {
```
Comment (charris) [inline numpy/_build_utils/src/apple_sgemv_fix.c]: Like so:
```
static inline int
using_mavericks(void)
{
```
also above.

**Comment by charris** [comment]
Still need C style fixups.

File: numpy/_build_utils/src/apple_sgemv_fix.c (hunk 5 above)
Comment (charris) [reply numpy/_build_utils/src/apple_sgemv_fix.c]: Strictly speaking, it should be `NPY_INLINE`, but now that we are requiring C99, `inline` is probably OK.

**Comment by bran206** [comment]
Sorry about the wait, got really busy.

**Comment by charris** [comment]
Let's give this a shot. Thanks @bran206 .

**Pull request merged**